signac/contrib/project.py

```diff
@@ -1474,7 +1470,7 @@ def temporary_project(self, dir=None):
             yield tmp_project
 
     @classmethod
-    def init_project(cls, *args, root=None, workspace=None, make_dir=True, **kwargs):
+    def init_project(cls, *args, root=None, **kwargs):
```

Comment:
I still think the choice to postpone some breaking changes until 3.0 is arbitrary and should be done in 2.0 instead. At this point, we're changing enough behavior that I expect user scripts to break. We might as well change everything we want to change, rather than maintaining the large and awkward API shims like in this function.

Reply:
That echos what we discussed in our dev meeting today, the more breakage we do now, the less we will have to break later/make a premature major version release when we could have set ourselves up better now.

Reply:
I don't think it's arbitrary, I think it's based on a reasonable guess of usage patterns based on empirical evidence, not to mention the fact that we're only going to be surfacing deprecation warnings for a short period of time because we're only now switching from DeprecationWarning to FutureWarning.

Having said that, I would be on board with just accepting that a major release involves breakage and not trying to maintain these patches. I agree that this is in line with what we discussed at the meeting. @csadorf what do you think? @mikemhenry could you add this as a topic for our next meeting?

Reply:
I agree with @vyasr that this is not arbitrary, but I also understand @bdice and @mikemhenry 's concern with respect to the overall pace of changes and this specific change. Version 1.0.0 of signac was released in Feb 2019, which means that we provided a very stable API for over **three years** now. That's a great accomplishment, but also comes with some responsibility to maintain that stability.

As a user of the _basic functions_* of a library that uses SemVer within the Python data ecosystem, I typically have the following expectations:

1. A major version change _will typically_ require a few high impact changes (adaptation effort on the order of hours) and will sometimes require substantial effort (weeks).
2. A minor version change _should not, but might_ require a few changes, but certainly with very low impact (minutes).
3. Any change required for a patch version upgrade is a bug/regression.

I would recommend that we move forward with all breaking changes for 2.0 without mitigation measures except where we estimate that the user effort for adaptation would be significantly more than 1 hour per project. For example, a breaking change to `init_project()` would be acceptable, because there are like only very few uses of this particular method for each project. A change to `.sp` or `.workspace` on the other hand would have a significantly larger impact.

In short: Let's move ahead with _all warranted_ breaking changes, but try to evaluate their potential impact and continue to support deprecated API with clear user instructions on how to upgrade as part of the warnings only for individual high impact changes.

*) Usually those shown as part of the quickstart examples and basic tutorials.

Reply:
RE: https://github.com/glotzerlab/signac/pull/714#discussion_r834971063
@cbkerr will be the community manager for the next meeting

Reply:
@csadorf What would you recommend for an API break that requires more than one hour to complete? There isn't anything bigger than a major version. Users can pin to signac 1.x for as long as they want, if they aren't willing to upgrade. I think it's in the long-term interest of the project to be open to API adjustments. Both freud 2.0 and HOOMD 2.0 / 3.0 involved major API breaks. I just don't see a convincing argument for 2.0 to be released with deprecated code paths that won't be removed until some indeterminate future time (we have no guarantee that a signac 3.0 will _ever_ exist, based on the slowing pace of development and high stability of the package).

Reply:
Yeah I'm starting to come around to this point of view as well, especially given 

> we have no guarantee that a signac 3.0 will ever exist, based on the slowing pace of core feature development and high stability of the package

Breaking changes are what major releases are for. While I still don't think our decisions were arbitrary, I don't know that there's a good enough reason to avoid the breakage going forward. That said... let's discuss this at the next dev meeting. I think that's a more efficient medium for this kind of discussion than async Github threads.

Reply:
> @csadorf What would you recommend for an API break that requires more than one hour for users to adapt? There isn't anything bigger than a major version.

@bdice I think I was not fully successful in conveying the message I tried to send. A major breaking change can absolutely require users to make adaptations that require more than one hour of effort. There is really no limit here. When I wrote about "orders of hours to weeks", I was really just talking about my personal expectations, this does not really have much to do with SemVer _per se_.

My main point is that this is not a black and white issue. SemVer allows you to release basically a completely new package that has an entirely different API (**zero compatibility!**) with a new major version. Is that "legitimate"? Yes. Would you be better off just releasing a new package under a different name? Maybe.

What I am advocating for is to _make any breaking change that we want_. Only for individual high impact changes, let's have a close look at wether it is a) necessary and b) could be easily mitigated. If the answers are a) yes and b) no, move ahead with the breaking change.

Reply:
I'm going to resolve this conversation and make a follow-up PR removing backwards compatibility layers. If there are specific ones that anyone feels strongly that we should keep, they can be discussed in that PR.